Notify the user when the file path contains leading or trailing spaces and fix the error message for invalid file names.

charles7668 opened the pull request:
close #31478

silverwind reviewed (approved):
Seems ok. Personally I would have tried to do it in frontend, but this is also fine and probably simpler than frontend stuff.

wxiaoguang commented on routers/web/repo/editor.go:
```diff
@@ -392,6 +392,14 @@ func EditFilePost(ctx *context.Context) {
+	for i, v := range treePath {
+		treePath[i] = strings.TrimSpace(v)
+	}
+	form.TreePath = strings.Join(treePath, "/")
```
Hmm, the path could be something like `/ / / /`, just press Space Slash Space Slash on the editor page 🤣 

Then the tree path becomes "////", which doesn't seem good.

silverwind replied:
Ah, we are already in `NewFilePost`, so likely above remark is invalid.

silverwind replied:
> Then the tree path becomes "////", which doesn't seem good.

I assume the backend would reject that? Seems like right behaviour to me.

wxiaoguang replied:
There are more edge cases: for example, what happens to `foo/ .. /bar`? Will it be rejected or become `foo/../bar` then become `bar`?

wxiaoguang replied:
Actually I think we could just warn the users on the UI, instead of doing too much.

The more we do, the more edge cases come.

silverwind replied:
> There are more edge cases: for example, what happens to `foo/ .. /bar`? Will it be rejected or become `foo/../bar` then become `bar`?

Same as entering `..` before and I hope we don't follow this reference in backend.

Not sure about warning, for me it seems clear-cut that a user would never want leading/trailing whitespace in the file or folder names and if they really do, they ought to use a git client.

charles7668 replied:
> Hmm, the path could be something like `/ / / /`, just press Space Slash Space Slash on the editor page 🤣
> 
> Then the tree path becomes "////", which doesn't seem good.

In this case, there is prevention by the front-end, we can't just type '/' in the path without any other character.

> There are more edge cases: for example, what happens to `foo/ .. /bar`? Will it be rejected or become `foo/../bar` then become `bar`?

In the current latest version, if the path is ' .. ' with leading or trailing spaces, it will cause some incorrect behavior because '..' with leading and trailing spaces is not allowed in the system.

For example, the test case 't1/ ../t2/test-1' will not be created, and the system will just respond with the same string.

Typing '..' might seem weird, but I believe that people who type '..' do it with meaningful intent.

charles7668 replied:
Preventing this on the front-end is a good solution, but currently, we can't see spaces in the UI.

This is the current UI display, but every t-n I have added some leading or trailing space, which is not being shown.
![image](https://github.com/go-gitea/gitea/assets/30816317/093a88ce-d051-483d-8fd8-be93ff6a8fab)

To prevent this on the front-end, we have two choices:

1. Change the display to show spaces. This will let people see what they have typed and display a warning.
2. Just show a warning (I think this is not ideal because people can't see the full typed path in the current UI).

wxiaoguang replied:
Tested with GitHub, I think GitHub also uses my suggestion: do not do too much for end users, only warn them.

wxiaoguang replied:
GitHub has another good behavior: if `../` is inputted, then it changes the path to parent, not sure whether it is implemented in Gitea's UI.

charles7668 replied:
> Tested with GitHub, I think GitHub also uses my suggestion: do not do too much for end users, only warn them.在 GitHub 上测试过，我认为 GitHub 也采用了我的建议：不要为最终用户做太多事情，只警告他们。

On GitHub, it shows a warning but does not take any action for users. This is my test: 't2  /  t3   /   t4  /test' 
![圖片](https://github.com/go-gitea/gitea/assets/30816317/28091f06-757d-464d-971e-bf46a263f6d9)
![圖片](https://github.com/go-gitea/gitea/assets/30816317/ef4c843c-4cd0-4b38-b598-2ffffb345ba6)

This problem is the same as #31478.

I'm not sure if we need to follow this behavior or not.

silverwind replied:
If GitHub has this warning, then I agree, we should implement it too.

delvh replied:
Both approaches sound good to me.
Both get an LGTM from me.

charles7668 replied:
![firefox_TrY7rDIQ4d](https://github.com/go-gitea/gitea/assets/30816317/671f22cd-81a9-4a82-bd40-17c7dece7b07)

I added some warnings for users similar to GitHub. 
However, I have kept backend processing for spaces. 
On GitHub, it only removes spaces from filenames, not from parent directories. Since Gitea can run on different operating systems, leading or trailing spaces may cause some problems. Therefore, I have retained space processing.

delvh commented:
@silverwind could you review again?

wxiaoguang commented:
1. IMO trimming spaces in backend is not right (well, just my personal option, not a blocker if most people like this behavior)
2. Coding style: `showElem` vs `.style.display = ...`
3. There are still bugs, for example:
    * `input` something like `foo/../bar` causes incorrect result.
    * `document.querySelector('.ui.warning.message.flash-message.flash-warning')` may query other "flash message" incorrectly but not the "space-related" one.

charles7668 commented:
> 1. IMO trimming spaces in backend is not right (well, just my personal option, not a blocker if most people like this behavior)
> 
>     2. Coding style: `showElem` vs `.style.display = ...`
> 
>     3. There are still bugs, for example:
>        
>        * `input` something like `foo/../bar` causes incorrect result.

How can this bug be reproduced? I tried it but couldn't reproduce it.
![firefox_V5c0V4tfCL](https://github.com/user-attachments/assets/db40e4f2-da40-4271-bec6-2cbd6b6d8ca4)


>        * `document.querySelector('.ui.warning.message.flash-message.flash-warning')` may query other "flash message" incorrectly but not the "space-related" one.
>        * the concern in https://github.com/go-gitea/gitea/pull/31507/files#r1657163784 is still not resolved, I do not think making `/ / / /` to `////` is right for backend path handling.

I will fix the other bug.

wxiaoguang commented:
> How can this bug be reproduced? I tried it but couldn't reproduce it.

Paste `foo/../bar`

charles7668 commented:
Fixed bugs

Regarding the backend problem, I think we might need more information.

Actually, I tested Gitea on both Windows and Linux servers by creating a `  test  /  test2  /test` file.

On the Linux server, this file can be created successfully.

On the Windows server, it returns the following page and fails to create the file:

![圖片](https://github.com/user-attachments/assets/626e92ec-fc59-4179-8c33-e70d1ee1c8c3)

In my opinion, for consistency, we need to maintain the same behavior across different OS servers.

wxiaoguang commented:
> On the Linux server, this file can be created successfully.
> On the Windows server, it returns the following page and fails to create the file:
> In my opinion, for consistency, we need to maintain the same behavior across different OS servers.


That's impossible, naturally Windows and Linux are different, eg: `the?file.txt`, `the*file.txt`, `the<file.txt`, `con`, `com1` .........

wxiaoguang commented:
> > On the Linux server, this file can be created successfully.
> > On the Windows server, it returns the following page and fails to create the file:
> > In my opinion, for consistency, we need to maintain the same behavior across different OS servers.
> 
> That's impossible, naturally Windows and Linux are different, eg: `the?file.txt`, `the*file.txt`, `the<file.txt`, `con`, `com1` .........

Personally I would suggest never to run Gitea server on Windows

diamante0018 commented:
Any update on this ? 🙏

lunny commented:
> As an alternative to trimming spaces, I fixed the error message when the file can't be created.
> 
> ![圖片](https://private-user-images.githubusercontent.com/30816317/350368683-eac316b0-8043-4b84-a676-1148aa5c8dfe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MDY3MDUsIm5iZiI6MTcyNTYwNjQwNSwicGF0aCI6Ii8zMDgxNjMxNy8zNTAzNjg2ODMtZWFjMzE2YjAtODA0My00Yjg0LWE2NzYtMTE0OGFhNWM4ZGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDA3MDY0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5OGQxNDU5NzEwYjI5ZTZlYmM1YmRmNWVmN2M4MDMwNWE2ZjJhODA0YjIwYzNhYjJlYzU1MmI5ZTMyODdiNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2sRwOaogLvST003zPiorF1KkdMONr7DEJxLQEnKjXDk)

Could you also change the title and resolve the conflicts?

charles7668 commented:
> > As an alternative to trimming spaces, I fixed the error message when the file can't be created.
> > ![圖片](https://private-user-images.githubusercontent.com/30816317/350368683-eac316b0-8043-4b84-a676-1148aa5c8dfe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MDY3MDUsIm5iZiI6MTcyNTYwNjQwNSwicGF0aCI6Ii8zMDgxNjMxNy8zNTAzNjg2ODMtZWFjMzE2YjAtODA0My00Yjg0LWE2NzYtMTE0OGFhNWM4ZGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDA3MDY0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5OGQxNDU5NzEwYjI5ZTZlYmM1YmRmNWVmN2M4MDMwNWE2ZjJhODA0YjIwYzNhYjJlYzU1MmI5ZTMyODdiNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2sRwOaogLvST003zPiorF1KkdMONr7DEJxLQEnKjXDk)
> 
> Could you also change the title and resolve the conflicts?

Updated and CI error seems not related

lunny commented:
It looks like I can still create a file name with trailing whitespace.
<img width="421" alt="图片" src="https://github.com/user-attachments/assets/e8baa5e8-e9b3-4c2b-835f-4f9abbfe3d52">

charles7668 commented:
> It looks like I can still create a file name with trailing whitespace. <img alt="图片" width="421" src="https://private-user-images.githubusercontent.com/81045/368392370-e8baa5e8-e9b3-4c2b-835f-4f9abbfe3d52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2Mjc5NDUsIm5iZiI6MTcyNjYyNzY0NSwicGF0aCI6Ii84MTA0NS8zNjgzOTIzNzAtZThiYWE1ZTgtZTliMy00YzJiLTgzNWYtNGY5YWJiZmUzZDUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDAyNDcyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllNDZjYTg5Y2ZkMzc4MjQ2NWU1NzI4ZDQ5Yjg1MzEwZWM4NGM0NTM0NDFiZWY3MTA0YTNmMDE4MjA1ZTU5NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f9BcAQhQq7FQ7s1_sRzMHd0tdjGY0M_D6yF7cA6IXoU">

Based on this comment, I have changed the approach to not prevent users from creating spaces in the file name. Instead, a warning message will be shown if there are trailing or leading spaces, and an error message will be displayed if the file cannot be created on the server OS.

I think whether or not to process spaces in the backend can be proposed for discussion.

> > On the Linux server, this file can be created successfully.
> > On the Windows server, it returns the following page and fails to create the file:
> > In my opinion, for consistency, we need to maintain the same behavior across different OS servers.
> 
> That's impossible, naturally Windows and Linux are different, eg: `the?file.txt`, `the*file.txt`, `the<file.txt`, `con`, `com1` .........

lunny commented:
> > It looks like I can still create a file name with trailing whitespace. <img alt="图片" width="421" src="https://private-user-images.githubusercontent.com/81045/368392370-e8baa5e8-e9b3-4c2b-835f-4f9abbfe3d52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2Mjc5NDUsIm5iZiI6MTcyNjYyNzY0NSwicGF0aCI6Ii84MTA0NS8zNjgzOTIzNzAtZThiYWE1ZTgtZTliMy00YzJiLTgzNWYtNGY5YWJiZmUzZDUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDAyNDcyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllNDZjYTg5Y2ZkMzc4MjQ2NWU1NzI4ZDQ5Yjg1MzEwZWM4NGM0NTM0NDFiZWY3MTA0YTNmMDE4MjA1ZTU5NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f9BcAQhQq7FQ7s1_sRzMHd0tdjGY0M_D6yF7cA6IXoU">
> 
> Based on this comment, I have changed the approach to not prevent users from creating spaces in the file name. Instead, a warning message will be shown if there are trailing or leading spaces, and an error message will be displayed if the file cannot be created on the server OS.
> 
> I think whether or not to process spaces in the backend can be proposed for discussion.
> 
> Leading or trailing spaces are allowed on a Linux server, but if using a Windows server, it will show an error message.
> 
> > > On the Linux server, this file can be created successfully.
> > > On the Windows server, it returns the following page and fails to create the file:
> > > In my opinion, for consistency, we need to maintain the same behavior across different OS servers.
> > 
> > 
> > That's impossible, naturally Windows and Linux are different, eg: `the?file.txt`, `the*file.txt`, `the<file.txt`, `con`, `com1` .........

Yes. But no warning or error was displayed.

charles7668 commented:
> > > It looks like I can still create a file name with trailing whitespace. <img alt="图片" width="421" src="https://private-user-images.githubusercontent.com/81045/368392370-e8baa5e8-e9b3-4c2b-835f-4f9abbfe3d52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2Mjc5NDUsIm5iZiI6MTcyNjYyNzY0NSwicGF0aCI6Ii84MTA0NS8zNjgzOTIzNzAtZThiYWE1ZTgtZTliMy00YzJiLTgzNWYtNGY5YWJiZmUzZDUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDAyNDcyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllNDZjYTg5Y2ZkMzc4MjQ2NWU1NzI4ZDQ5Yjg1MzEwZWM4NGM0NTM0NDFiZWY3MTA0YTNmMDE4MjA1ZTU5NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f9BcAQhQq7FQ7s1_sRzMHd0tdjGY0M_D6yF7cA6IXoU">
> > 
> > 
> > Based on this comment, I have changed the approach to not prevent users from creating spaces in the file name. Instead, a warning message will be shown if there are trailing or leading spaces, and an error message will be displayed if the file cannot be created on the server OS.
> > I think whether or not to process spaces in the backend can be proposed for discussion.
> > Leading or trailing spaces are allowed on a Linux server, but if using a Windows server, it will show an error message.
> > > > On the Linux server, this file can be created successfully.
> > > > On the Windows server, it returns the following page and fails to create the file:
> > > > In my opinion, for consistency, we need to maintain the same behavior across different OS servers.
> > > 
> > > 
> > > That's impossible, naturally Windows and Linux are different, eg: `the?file.txt`, `the*file.txt`, `the<file.txt`, `con`, `com1` .........
> 
> Yes. But no warning or error was displayed.

I referred to GitHub's behavior, which only checks when typing '/'. I have now fixed it to check after every typing.

lunny commented:
I think this can be kept v1.23 only.
